src/PantherActions.php

```diff
@@ -278,49 +272,53 @@ protected static function submitForm(string $selector): void
 
     protected static function findFormField(string $fieldText, string $legend = null, string $contextSelector = null): Crawler
     {
-        $xpath = <<<'XPATH'
-            [
-                contains(concat(' ', normalize-space(string(.)), ' '), %1$s) or
-                contains(concat(' ', normalize-space(string(@value)), ' '), %1$s) or
-                @id=%2$s or
-                @name=%2$s or
-                @placeholder=%2$s
-            ]
-            XPATH;
-
+        // create crawler, optionally start by context
         $crawler = self::crawlerBySelector($contextSelector);
 
+        // if legend is given, search for that and get its fieldset
         if ($legend !== null) {
-            $crawler = $crawler
-                ->filterXPath(
-                    sprintf(
-                        '//legend' . $xpath . '/ancestor::fieldset',
-                        Crawler::xpathLiteral(' ' . $legend . ' '),
-                        Crawler::xpathLiteral($legend),
-                    )
-                )
-            ;
+            $crawler = $crawler->filterXPath(
+                self::formFieldXpath('descendant-or-self::legend', $legend) . '/ancestor::fieldset'
+            );
         }
 
-        foreach (['//input', '//label', '//select', '//option'] as $tag) {
-            $parts[] = $tag . $xpath;
+        // find form labels or elements with the given `$fieldText` as id, name, text, or placeholder
+        $field = $crawler->filterXPath(
+            implode(' | ', array_map(
+                static fn (string $tag): string => self::formFieldXpath("descendant-or-self::{$tag}", $fieldText),
+                ['label', 'input', 'select', 'option', 'textarea'],
+            ))
+        );
+
+        $foundFieldTag = $field->getElement(0)?->getTagName();
+
+        // Find input field connected to label.
+        if ($foundFieldTag === 'label') {
+            if ($id = trim($field->attr('for'))) {
+                // filter by id
+                $field = self::crawler()->filter("#{$id}");
+            } else {
+                // search for input types in descendants
+                $field = $field->filterXPath(
+                    implode(' | ', array_map(
+                        static fn (string $tag): string => "descendant::{$tag}",
+                        ['input | select | option | textarea'],
+                    ))
+                );
+            }
         }
 
-        $field = $crawler
-            ->filterXPath(
-                sprintf(
-                    implode(' | ', $parts),
-                    Crawler::xpathLiteral(' ' . $fieldText . ' '),
-                    Crawler::xpathLiteral($fieldText),
-                )
-            )
-        ;
+        if ($field->count() === 0) {
```

Comment:
Prevents the somewhat cryptic messages provided by the remote web driver. Instead a clear error message is now shown, optionally with your context selector. And when searched by label, that information is given too.